File: backend-crud/Backend/EmployeeController.fs
module EmployeeController

open Microsoft.AspNetCore.Http
open Newtonsoft.Json
open System.Collections.Generic
open System.IO
open Microsoft.FSharp.Control
open Model
open Validation

open RebelSoftware.LoggingService
open RebelSoftware.MessageQueueService
open RebelSoftware.HttpService.Http

//use this function to return a success value without microservices running
let create_dummy (httpContext:HttpContext) = 
    httpContext.Response.Headers.["Access-Control-Allow-Origin"] <- Microsoft.Extensions.Primitives.StringValues("*")
    httpContext.Response.WriteAsync("{Status:'Success'}")



//todo: look up RabbitMQ prod guidelines. (this code is based on the C# tutorial, which is not the best practice for prod)
//todo: return error status to client if write to Rabbit MQ failed, or if microservice failed, or isn't running
//todo: shared logging layer between backend and microservices?
let create (logger:Logging.Logger) (messageQueuer:MessageQueueing.MessageQueuer) (httpServer:HttpServer) (employeeValidator:EmployeeValidator) =
    let getResponseStr requestJsonString opResult = 
        match opResult.ValidationResult = ValidationResults.Success with 
            | true ->
                messageQueuer.WriteMessageAndGetResponse requestJsonString
            | false -> 
                opResult :> obj
                |> httpServer.SerializeToJson

    try
        //UNIT TEST: 
        //  * when body contains valid Employee object, we write the response from the message queue. 
        //  * when body contains validation errors, we write the OperationResult w/ expected validation errors to the response 
        //  * when body contains invalid JSON, we write an OperationResult w/ UnknownError to the response, and call logger.LogException
        let requestJsonString = httpServer.ReadRequestBody ()
        requestJsonString
        |> httpServer.DeserializeEmployeeFromJson 
        |> employeeValidator.ValidateEmployee
        |> (getResponseStr requestJsonString)
        |> httpServer.WriteHttpResponse
    with
    | ex -> 
        logger.LogException ex |> ignore
        { ValidationResult = ValidationResults.UnknownError }
        |> JsonConvert.SerializeObject
        |> httpServer.WriteHttpResponse 

    

File: backend-crud-CSharp/backend-crud-CSharp/EmployeeController.cs
using System;

using RebelSoftware.MessageQueueService;
using RebelSoftware.SerializationService;
using RebelSoftware.HttpService;

using RebelSoftware.Logging;

namespace backend_crud_CSharp
{
    public class EmployeeController
    {
        private ILoggingService _logger;
        private MessageQueueing.MessageQueuer _messageQueuer;
        private Serialization.SerializationService<Model.Employee> _serializationService;
        private Validation.EmployeeValidator _employeeValidator;
        private IHttpService _httpService;
        public EmployeeController(ILoggingService logger, MessageQueueing.MessageQueuer messageQueuer, Serialization.SerializationService<Model.Employee> serializationService, IHttpService httpService, Validation.EmployeeValidator employeeValidator)
        {
            _logger = logger;
            _messageQueuer = messageQueuer;
            _serializationService = serializationService;
            _employeeValidator = employeeValidator;
            _httpService = httpService;
        }

        private string GetResponseString(string requestJsonString, Model.OperationResult operationResult)
        {
            string retVal = null;

            if(operationResult.ValidationResult == Model.ValidationResults.Success)
            {
                retVal = _messageQueuer.WriteMessageAndGetResponse.Invoke(requestJsonString);
            }
            else
            {
                var objOperationResult = (object)operationResult;
                retVal = _serializationService.SerializeToJson.Invoke(objOperationResult);
            }
            
            return retVal;
        }

        public void Create()
        {
            try
            {
                var requestJsonString = _httpService.ReadRequestBody();
                var employee = _serializationService.DeserializeFromJson.Invoke(requestJsonString);
                var operationResult =  _employeeValidator.ValidateEmployee.Invoke(employee);
                var responseStr = GetResponseString(requestJsonString, operationResult);
                _httpService.WriteHttpResponse(responseStr);
            }
            catch(Exception ex)
            {
                _logger.LogException(ex);
                var operationResult = new Model.OperationResult(Model.ValidationResults.UnknownError);
                var responseStr = _serializationService.SerializeToJson.Invoke(operationResult);
                _httpService.WriteHttpResponse(responseStr);
            }
        }
    }
}